tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "mococrw/symmetric_crypto.h"
+
+using namespace mococrw;
+
+using EncrytDecryptTestData = std::tuple<SymmetricCipherMode, std::string>;
+
+static const std::vector<SymmetricCipherMode> AllSupportedCIpherModesToTest{
+        SymmetricCipherMode::CBC,
+        SymmetricCipherMode::GCM,
+        SymmetricCipherMode::CTR
+};
+
+class SymmetricCipherBase
+{
+public:
+    void SetUp()  {
+        _secretKey = utility::randomBytes(256 / 8);
+
+        std::string tmp;
+        for (auto i = 0; i < 1024 + 17; i++) {
+            tmp += "a string which will be concatenated multiple times-";
+        }
+        _plaintext = {tmp.begin(), tmp.end()};
+    }
+
+    std::vector<uint8_t> _secretKey;
+    std::vector<uint8_t> _plaintext;
+};
+
+class SymmetricCipherTest : public SymmetricCipherBase, public testing::TestWithParam<EncrytDecryptTestData>
+{
+    void SetUp() override  {
+        SymmetricCipherBase::SetUp();
+    }
+};
+
+class SymmetricCipherAdvancedTest : public SymmetricCipherBase, public testing::TestWithParam<SymmetricCipherMode>
+{
+    void SetUp() override  {
+        SymmetricCipherBase::SetUp();
+    }
+};
+
+
+TEST_P(SymmetricCipherTest, EncryptDecryptDifferentPlaintextLength)
+{
+    auto operationMode = std::get<0>(GetParam());
+    auto plaintextString = std::get<1>(GetParam());
+    std::vector<uint8_t> plaintext{plaintextString.begin(), plaintextString.end()};
+
+    // When: encrypt and then decrypt the plaintext
+    auto encryptoBuilder = AESCipherBuilder{operationMode, _secretKey};
+    std::shared_ptr<SymmetricCipher> encryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        encryptor = encryptoBuilder.buildAuthenticatedEncryptor();
+    }
+    else {
+        encryptor = encryptoBuilder.buildEncryptor();
+    }
+
+    encryptor->update(plaintext);
+    auto ciphertext = encryptor->finish();
+    auto iv = encryptor->getIV();
+
+    std::vector<uint8_t> tag;
+    auto authenticatedEncryptor = dynamic_cast<AuthenticatedEncryptionI*>(encryptor.get());
```

Comment:
I don't know if we want to encourage this style.

Reply:
it is a trade-off between maintainability of the tests and readability. I can add a comment explaining that it is not an intended use-case. What would you suggest?

Reply:
ok